Updating graph engine to apply constraints

aaronaddleman opened the pull request:
It was found that the PngEngine was not utilizing the GraphConstants.MaxWidth and GraphConstants.MaxHeight. Applying a validation step in the createImage of the DefaultGraphEngine so that all requests receive this clamp of values will help reduce the chance of a out of memory error. This applies to requests with valid or invalid queries that have dimensions that are out of bounds for the GraphConstants values defined.

aaronaddleman committed:
Updating graph engine to apply constraints

It was found that the userError and systemError were not utilizing the GraphConstants.MaxWidth and GraphConstants.MaxHeight. Applying a validation step in the createImage of the DefaultGraphEngine so that all requests receive this clamp of values will help reduce the chance of a out of memory error. This applies to requests with valid or invalid queries that have dimensions that are out of bounds for the GraphConstants values defined.

brharrington commented on atlas-chart/src/main/scala/com/netflix/atlas/chart/util/PngImage.scala:
```diff
@@ -87,12 +89,22 @@ object PngImage {
     val systemErrorRed = new Color(0xF8, 0x20, 0x00)
-    error(imgText, width, height, "SYSTEM ERROR:", Color.WHITE, systemErrorRed)
+    val clampedWidth = math.min(width, GraphConstants.MaxWidth)
+    val clampedHeight = math.min(height, GraphConstants.MaxHeight)
```
I would suggest moving this to the `error` method so it covers all the cases in a single place.

aaronaddleman replied:
Thanks for pointing this out! Will do

aaronaddleman replied:
This has ben moved to error()

brharrington commented on atlas-chart/src/main/scala/com/netflix/atlas/chart/util/PngImage.scala:
```diff
@@ -87,12 +89,22 @@ object PngImage {
+    
+    val dimensionWarning = if (width > GraphConstants.MaxWidth || height > GraphConstants.MaxHeight) {
+      s" Image dimensions clamped to ${clampedWidth}x${clampedHeight} (max: ${GraphConstants.MaxWidth}x${GraphConstants.MaxHeight})."
+    } else ""
```
Not sure we need to append this detail to the error images. They will see it as a warning on the regular graph once they fix whatever error is being shown. Adding details about restricting the image size to the error could just cause confusion and distract from the issue that led to the error.

aaronaddleman replied:
Yah I can take it out if you like. My thoughts were about surfacing what value was expected. No problem, will remove.

aaronaddleman replied:
I think this is now resolved

aaronaddleman committed:
Adding tests for the DefaultGraphEngineSuite and GraphConstantsSuite

aaronaddleman committed:
Move clampedHeight and clampedWith to error for general validation on all errors.

aaronaddleman committed:
Formatting cleanup

aaronaddleman committed:
Formatting cleanup

aaronaddleman commented:
Here are examples of the current testing.

- bad query with height value beyond GraphConstant.maxHeight
<img width="500" height="1000" alt="test badQuery 1" src="https://github.com/user-attachments/assets/520a75f9-2b87-4aa3-a301-f3d498b4a3c3" />

- good query with height value beyond GraphConstant.maxHeight
<img width="586" height="1168" alt="test random 1" src="https://github.com/user-attachments/assets/5b2b1dd4-f470-48ed-979b-559c376e1194" />

aaronaddleman committed:
Add a total of 2 benchmark tests

pngImageGeneration benchmarks the DefaultGraphEngine.createImage() with default values for the image

oversizedImageGeneration benchmarks the DefaultGraphEngine.createImage() with values that are beyond the GraphConstants.MaxWidth and GraphConstants.MaxHeight

aaronaddleman committed:
Add tests for validation

These tests are using the GraphConstants.validate and asserting values which are not meeting the required value ranges invoke the adding of the validation clamped values which matc the GraphConstants default values.

aaronaddleman committed:
Minor updates to dimension testing

brharrington commented on atlas-chart/src/main/scala/com/netflix/atlas/chart/util/PngImage.scala:
```diff
@@ -87,7 +89,10 @@ object PngImage {
     val userErrorYellow = new Color(0xFF, 0xCF, 0x00)
-    error(imgText, width, height, "USER ERROR:", Color.BLACK, userErrorYellow)
+    val clampedWidth = math.min(width, GraphConstants.MaxWidth)
+    val clampedHeight = math.min(height, GraphConstants.MaxHeight)
```
This will be done when calling error, so shouldn't need to be repeated here

aaronaddleman replied:
Ok I will remove this

aaronaddleman committed:
Remove duplicate code effect in userError
